devito/petsc/solve.py

```diff
@@ -105,3 +115,51 @@ def _(expr, target):
 @remove_target.register(Derivative)
 def _(expr, target):
     return 0 if expr.has(target) else expr
+
+
+def centre_stencil(eqn, target):
+    """
+    Extract the centre stencil from equation.
+    This function assumes that the core stencil, which is the stencil from
+    which the centre stencil will be extracted, is located on the RHS of
+    the input equation (first argument).
+
+    NOTE: At the point of entry, the time derivatives are likely evaluated, but
+    not the spatial derivatives. This necessitates evaluating 'eqn'
+    before deriving the centre stencil. By doing so, we ensure that
+    all derivatives, including spatial derivatives, are correctly accounted for
+    in the centre stencil.
+    """
+    centre = extract_centre(eqn.evaluate.rhs, target)
+    return centre
+
+
+@singledispatch
+def extract_centre(expr, target):
+    return expr if expr == target else 0
+
+
+@extract_centre.register(Add)
+@extract_centre.register(EvalDerivative)
+def _(expr, target):
+    if not expr.has(target):
+        return 0
+
+    args = [extract_centre(a, target) for a in expr.args]
+    return expr.func(*args, evaluate=False)
+
+
+@extract_centre.register(Mul)
+def _(expr, target):
```

Comment:
Could this be consolidated with the `Add`/`EvalDerivative` dispatch? Seems like the only difference is the `if not a.has(target)`

Reply:
I don't think so